### PR TITLE
feat(python)!: Improve some error types and messages

### PR DESCRIPTION
**Pull request by aminalaee**
Partially addresses https://github.com/pola-rs/polars/issues/10421

Changes:
* Improve error messages to be in accordance with the guidelines (see linked issue).
* Fix the error types of some functions. This could be considered a breaking change.

### CODE REVIEW

File: py-polars/polars/convert.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ def from_dicts(
 
     """
     if not data and not (schema or schema_overrides):
-        raise NoDataError("No rows. Cannot infer schema.")
+        raise NoDataError("No rows. Cannot infer schema")
```
Comment (ritchie46) [inline py-polars/polars/convert.py:176]: Should also remove the leading Capital.

File: py-polars/polars/utils/_parse_expr_input.py [hunk 2]
```diff
@@ -132,7 +132,7 @@ def parse_as_expression(
     else:
         raise TypeError(
             f"did not expect value {input!r} of type {type(input)}, maybe disambiguate with"
-            " pl.lit or pl.col"
+            "\n\npl.lit or pl.col"
```
Comment (stinodego) [inline py-polars/polars/utils/_parse_expr_input.py:135]: Your regexes are definitely leading to some false positives.

**Review by stinodego — CHANGES_REQUESTED** [review]
Thanks for the effort! Though I definitely see some issues still - messages that start with a capital letter, or mistakenly added newlines.

I'll put this on draft, just mark it ready-to-review when you've made a pass to verify your changes!

**Comment by alexander-beedie** [comment]
Nice one! I was going to look at this but looks like you're off to a good start - go for it :))

File: py-polars/polars/dataframe/frame.py [hunk 3]
```diff
@@ -840,7 +840,7 @@ def _read_parquet(
             else:
                 raise ValueError(
                     "cannot use glob patterns and integer based projection as `columns`"
-                    " argument; Use columns: List[str]"
+                    "\n\nargument; Use columns: List[str]"
```
Comment (mcrumiller) [inline py-polars/polars/dataframe/frame.py:843]: I think you put the newlines in the wrong place, should it not go after "argument"? Same with 952 below.

**Comment by mcrumiller** [comment]
Out of curiosity, would we benefit from an error-helper function, or is this too much indirection? Something like

```python
pl.exceptions.raise(
    "ValueError",
    header="polars can not currently guarantee zero-copy conversion from Arrow for categorical columns",
    desc="Set `allow_copy=True` or cast categorical columns to string first."
)
```
Which could then ensure proper formatting, and insert newlines where appropriate?

File: py-polars/polars/dataframe/frame.py (hunk 3 above)
Comment (mcrumiller) [reply py-polars/polars/dataframe/frame.py:843]: Edit: I see this was mentioned ago, and you regex'd this which makes sense. Carry on!

**Comment by alexander-beedie** [comment]
> Out of curiosity, would we benefit from an error-helper function, or is this too much indirection? Something like

Perhaps, but it would find its way into the reported stack which isn't ideal.

**Comment by mcrumiller** [comment]
@alexander-beedie there are ways to modify the traceback with the [`traceback`](https://docs.python.org/3/library/traceback.html) module, Jinja is well-known for doing this. But it's probably more effort than it's worth, especially if we want to be compatible on multiple versions.

**Comment by alexander-beedie** [comment]
> @alexander-beedie there are ways to modify the traceback with the [`traceback`](https://docs.python.org/3/library/traceback.html) module, Jinja is well-known for doing this. But it's probably more effort than it's worth, especially if we want to be compatible on multiple versions.

Indeed; doing "clever" things with your exceptions IN your exception handling code is something I'm not a big fan of ;)

**Review by stinodego — APPROVED** [review]
Nice improvement!

I did have to fix a few false positives, and there are many messages that still do not conform to the guidelines. Plus the same improvements should be done on the Rust side. So the linked issue remains open for now.

If you wish to continue to work on the issue, I would advise a more concentrated approach: pick a module and make sure all error messages are perfect. That way, it's easier for us to review.

**Comment by mcrumiller** [comment]
@aminalaee 55 files changed! Good job here. We all encounter error messages every day so this is a nice improvement :)

**Comment by aminalaee** [comment]
Thanks for the review and kind words.
Yes I would love to keep working on the issue on both Python and Rust side and as you said it is better to go module by module as there are cases that are harder to find. For example I saw this by going through the files:


```py
msg = “Invalid operation”
raise ValueError(msg)
```

**Pull request merged**